src/llm/servable_initializer.cpp

```diff
@@ -49,11 +51,91 @@
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+static std::pair<std::optional<std::string>, std::optional<std::string>> getBosAndEosTokenFromTokenizerVocab(const std::shared_ptr<ov::genai::Tokenizer>& tokenizer) {
+    auto vocab = tokenizer->get_vocab();
+    SPDLOG_TRACE("Tokenizer vocab size: {}", vocab.size());
+    auto bosTokenId = tokenizer->get_bos_token_id();
+    auto eosTokenId = tokenizer->get_eos_token_id();
+    std::optional<std::string> bosToken;
+    std::optional<std::string> eosToken;
+    // since tokenizer get_bos_token does not work for gguf we will search in map by value
+    for (const auto& [token, id] : vocab) {
+        // SPDLOG_TRACE("Tokenizer vocab token: {}, id: {}", token, id);
+        if (id == bosTokenId) {
+            bosToken = token;
+        }
+        if (id == eosTokenId) {
+            eosToken = token;
+        }
+    }
+    return std::make_pair(bosToken, eosToken);
+}
+
 void GenAiServableInitializer::loadPyTemplateProcessor(std::shared_ptr<GenAiServableProperties> properties, const std::string& chatTemplateDirectory) {
     py::gil_scoped_acquire acquire;
     try {
-        auto locals = py::dict("templates_directory"_a = chatTemplateDirectory);
+        bool isGGUFModel = checkIfGGUFModel(chatTemplateDirectory);
+        // we need to pass tokenizer template and bos/eos tokens to python code
+        // if we have GGUF model, we will use them to create a template object
+        std::string tokenizerTemplate = properties->tokenizer.get_chat_template();
+        std::string tokenizerBosToken = properties->tokenizer.get_bos_token();
+        std::string tokenizerEosToken = properties->tokenizer.get_eos_token();
+        // time measure following if statement
+        std::chrono::steady_clock::time_point start = std::chrono::steady_clock::now();
+        // Workaround for CVS-172426
+        if (isGGUFModel && (tokenizerBosToken.empty() || tokenizerEosToken.empty())) {
+            // if tokenizer bos/eos tokens are empty, we will try to get them from tokenizer vocab
+            std::pair<std::optional<std::string>, std::optional<std::string>> tokens;
+            tokens = getBosAndEosTokenFromTokenizerVocab(std::make_shared<ov::genai::Tokenizer>(properties->tokenizer));
+            SPDLOG_TRACE("Tokens bos: {}, eos: {}", tokens.first.has_value() ? tokens.first.value() : "None", tokens.second.has_value() ? tokens.second.value() : "None");
+            if (tokens.first.has_value()) {
+                tokenizerBosToken = tokens.first.value();
+            }
+            if (tokens.second.has_value()) {
+                tokenizerEosToken = tokens.second.value();
+            }
+        }
+        std::chrono::steady_clock::time_point end = std::chrono::steady_clock::now();
+        SPDLOG_TRACE("Time to get bos/eos tokens from tokenizer: {} ms", std::chrono::duration_cast<std::chrono::microseconds>(end - start).count() / 1000.0);
```

Comment:
is it debug or will it stay?

Reply:
it may be usefull to know what's the impact of WA on different models. I don;t see harm in keeping this log.